src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,234 @@
+// featbin/wav-reverberate.cc
+
+// Copyright 2015  Tom Ko
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "feat/wave-reader.h"
+#include "feat/signal.h"
+
+namespace kaldi {
+
+void AddVectorsOfUnequalLength(const Vector<BaseFloat> &filter, Vector<BaseFloat> *signal) {
+  for (int po = 0; po < signal->Dim(); po += filter.Dim()) {
+    int block_length = filter.Dim();
+    if (signal->Dim() - po < block_length) block_length = signal->Dim() - po;
+    signal->Range(po, block_length).AddVec(1.0, filter.Range(0, block_length));
+  }
+}
+
+BaseFloat MaxAbsolute(const Vector<BaseFloat> &vector) {
+  return std::max(std::abs(vector.Max()), std::abs(vector.Min()));
+}
+
+BaseFloat ComputeEarlyReverbEnergy(const Vector<BaseFloat> &rir, const Vector<BaseFloat> &signal,
+                                   BaseFloat samp_freq) {
+  int peak_index = 0;
+  rir.Max(&peak_index);
+  KALDI_VLOG(1) << "peak index is " << peak_index;
+
+  const float sec_before_peak = 0.001;
+  const float sec_after_peak = 0.05;
+  int early_rir_start_index = peak_index - sec_before_peak * samp_freq;
+  int early_rir_end_index = peak_index + sec_after_peak * samp_freq;
+  if (early_rir_start_index < 0) early_rir_start_index = 0;
+  if (early_rir_end_index > rir.Dim()) early_rir_end_index = rir.Dim();
+
+  int duration = early_rir_end_index - early_rir_start_index;
+  Vector<BaseFloat> early_rir(rir.Range(early_rir_start_index, duration));
+  Vector<BaseFloat> early_reverb(signal);
+  BlockConvolveSignals(early_rir, &early_reverb);
+
+  // compute the energy
+  return VecVec(early_reverb, early_reverb) / early_reverb.Dim();
+}
+
+void DoReverberation(const Vector<BaseFloat> &rir, Vector<BaseFloat> *noise,
+                        Vector<BaseFloat> *signal, BaseFloat samp_freq,
+                        BaseFloat snr_db) {
+  if (noise->Dim()) {
+    float input_power = ComputeEarlyReverbEnergy(rir, *signal, samp_freq);
+    float noise_power = VecVec(*noise, *noise) / noise->Dim();
+    float scale_factor = sqrt(pow(10, -snr_db / 10) * input_power / noise_power);
+    noise->Scale(scale_factor);
+    KALDI_VLOG(1) << "scale_factor of channel: " << scale_factor << " for SNR db " << snr_db;
```

Comment:
KALDI_VLOG(1) << "Noise signal is being scaled with the value " << scale_factor << " to generate output with SNR  " << snr_db <<  "db\n";